org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/UnitTestSupport.java

```diff
@@ -0,0 +1,776 @@
+        launchArguments.append("-ea").append("\n");
+      }
+
+      if(args.getArgLine() != null) {
+        launchArguments.append(args.getArgLine()).append("\n");
+      }
+      if(args.getSystemPropertyVariables() != null) {
+        args.getSystemPropertyVariables().entrySet().forEach(
+            e -> launchArguments.append("-D").append(e.getKey()).append("=").append(e.getValue()).append("\n"));
+      }
+
+      copy.setAttribute(LAUNCH_CONFIG_VM_ARGUMENTS, launchArguments.toString());
+
+      try {
+        if(args.getWorkingDirectory() != null && !Files.isSameFile(project.getLocation().toPath().toAbsolutePath(),
+            args.getWorkingDirectory().toPath())) {
+          copy.setAttribute(LAUNCH_CONFIG_WORKING_DIRECTORY, args.getWorkingDirectory().getAbsolutePath());
+        } else {
+          copy.setAttribute(LAUNCH_CONFIG_WORKING_DIRECTORY, (String) null);
+        }
+      } catch(IOException ex) {
+        log.error(ex.getMessage(), ex);
+      }
+
+      if(args.getEnvironmentVariables() != null) {
+        copy.setAttribute(LAUNCH_CONFIG_ENVIRONMENT_VARIABLES, args.getEnvironmentVariables());
+      }
+
+      copy.doSave();
+    }
+
+    private static IMavenProjectFacade getMavenProjectFacade(IJavaProject project) {
+      return MavenPlugin.getMavenProjectRegistry().getProject(project.getProject());
+    }
+
+    /**
+     * Get the launch arguments
+     * 
+     * @param configuration the configuration
+     * @param facade the maven project facade
+     * @param monitor the progress monitor
+     * @return the launch arguments
+     * @throws CoreException
+     */
+    private TestLaunchArguments getTestLaunchArguments(ILaunchConfiguration configuration, IMavenProjectFacade facade,
+        IProgressMonitor monitor) throws CoreException {
+
+      String testClass = configuration.getAttribute(LAUNCH_CONFIG_MAIN_CLASS, "");
+
+      MavenProject mavenProject = facade.getMavenProject();
+
+      // find test executions
+      List<MojoExecution> executions = new ArrayList<>();
+      for(ExecutionId id : TEST_EXECUTIONS) {
+        executions.addAll(facade.getMojoExecutions(id.getGroupId(), id.getArtifactId(), monitor, id.getGoal()));
+      }
+
+      // find which plugin executions will launch the test
+      List<MojoExecution> handlers = executions.stream()
+          .filter(e -> isTestHandledByPlugin(facade, monitor, e, testClass)).toList();
+
+      // only one execution is expected here but if more only the first one is used
+      for(MojoExecution execution : handlers) {
+        return getTestLaunchArguments(mavenProject, execution, monitor);
+      }
+
+      return null;
+    }
+
+    /**
+     * Execute the prerequisites mojos
+     * 
+     * @param configuration the configuration
+     * @param facade the maven project facade
+     * @param monitor the progress monitor
+     */
+    private void loadPrerequisites(ILaunchConfiguration configuration, IMavenProjectFacade facade,
```

Comment:
Can't/shouldn't the list of mojos executed before the test-plugin be derived from the computed Maven lifecycle?

Reply:
This one is a bit more complex as i think it can greatly impact performance. I'm afraid executing all plugins available in the maven lifecycle until this phase without knowing which ones are effectively producing properties is risky.
The only plugin provided by the maven team can safely be handled by default but about the others ....i'm not sure.
On the other hand providing the list of plugins to execute using a property feals more like a dirty hack than a clean solution.

If you have a way to know if a plugin generate some properties or a cleaner way to provide the list of plugin to execute it would be great as i plan to use the same kind of loading in another PR to be able to provide values for --patch-module directives from the compiler plugin to eclipse .classpath file in the fututre.

At this point i only can think of other ways like:
- providing a PR to maven surefire/failsafe/compiler plugin to provide the same kind of feature as the property but i'm skeptical the maven team will accept this PR more dedicated to IDE/Eclipse than maven itself
- providing some warning markers and some potential "quick fix" into the pom warning the user that plugins aren't executed until some configuration (property/project's preference) is set. (I'm completly clueless about how to do that but it would be cool but also time consuming)

What do you (or any other interested ) think about that?

Reply:
Hello @HannesWell,
After some research it seems the best way to handle this may be the lifecycle-metadata-mapping.
Currently a plugin may have the following configuration in the mapping:

ignore
execute(runOnConfiguration/runOnIncremental)
generator

So i propose to add a boolean "runOnLaunchConfiguration" or something more generic to fit the compiler case like "provideProperties" to the execute element to mark all plugins that need to be executed before setting up a launch configuration. 

About the "generator" case, a new method can be added to org.eclipse.m2e.core.project.configurator.AbstractProjectConfigurator :

`public void configureLaunchConfiguration(LaunchConfigurationRequest request, IProgressMonitor monitor) throws CoreException`

or something like

`public void loadPropertiesInContext(??, IProgressMonitor monitor) throws CoreException`

What do you think about that ?

Reply:
I think leveraging the lifecycle-mapping-metadata is the the better approach.
It would be ideal if we could re-use the existing metadata, but I assume it doesn't fit well in all cases, does it?
If new arguments are added, it should as general as possible and as specific as necessary.

But from what you have described enhancing the https://github.com/codehaus-plexus/plexus-build-api would maybe be another option?

Maybe @laeubi can help here too?

Reply:
In order to simplify the review of this big change, I think it would be good to move the resolution of variables created by other plugins into a separate PR.
Would that be ok?

Reply:
ok let's continue this discussion in another PR ;)